src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3413,29 +3413,31 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           // repeat vararg as often as needed, remove by-name
           val formals = formalTypes(paramTypes, argslen)
 
-          /* Try packing all arguments into a Tuple and apply `fun`
-           * to that. This is the last thing which is tried (after
-           * default arguments)
+          /* Try packing all arguments into a Tuple and apply `fun` to that.
+           * This is the last thing which is tried (after default arguments).
            */
-          def tryTupleApply: Tree = {
-            if (eligibleForTupleConversion(paramTypes, argslen) && !phase.erasedTypes) {
+          def tryTupleApply: Tree =
+            if (phase.erasedTypes || !eligibleForTupleConversion(paramTypes, argslen)) EmptyTree
+            else {
               val tupleArgs = List(atPos(tree.pos.makeTransparent)(gen.mkTuple(args)))
               // expected one argument, but got 0 or >1 ==>  try applying to tuple
               // the inner "doTypedApply" does "extractUndetparams" => restore when it fails
               val savedUndetparams = context.undetparams
-              silent(_.doTypedApply(tree, fun, tupleArgs, mode, pt)) map { t =>
-                // Depending on user options, may warn or error here if
-                // a Unit or tuple was inserted.
-                val keepTree = (
-                     !mode.typingExprNotFun // why? introduced in 4e488a60, doc welcome
-                  || t.symbol == null       // ditto
-                  || checkValidAdaptation(t, args)
-                )
+              // May warn or error here if a Unit or tuple was inserted.
+              def validate(t: Tree): Tree = {
```

Comment:
This goes in the right direction in making the check more explicit. Can we push it a bit more and avoid future reorderings of the conditions?
```scala
// May warn or error if a Unit or tuple was inserted.
def validate(t: Tree): Tree = {
  val invalidAdaptation = t.symbol != null && !checkValidAdaptation(t, args) // regardless of typer's mode
  if (invalidAdaptation && mode.typingExprNotFun) EmptyTree // only bail if we're typing an expression (and not inside another application)
  else t
}
```